### ISSUE TITLE
CPU usage regression with SDLGL-PP

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1245]
Before 706793b5e27a012b71fcb02d208d225eef32fb47 CPU usage of openMSX was around 10-12%, but after that commit, it became irregular, with large amounts of time being 70-100%.
This is probably also the issue reported here: https://www.msx.org/forum/msx-talk/openmsx/energy-usage

SDL renderer performance was not affected, before and after this patch, it's around 33% on my system (3x scaling instead of 4x, though).

**Comment by MBilderbeek** [comment]
This is the output of perf top at a time the CPU load is high according to normal top:
```
Samples: 975K of event 'cycles', 4000 Hz, Event count (approx.): 54125531753 lost: 0/0 drop: 0/0
Overhead  Shared Object                       Symbol
   7,33%  libnvidia-glcore.so.440.82          [.] 0x0000000000f99565
   6,34%  [kernel]                            [k] apic_timer_interrupt
   5,76%  [kernel]                            [k] entry_SYSCALL_64
   2,92%  [kernel]                            [k] syscall_return_via_sysret
   2,77%  [kernel]                            [k] irq_entries_start
   2,76%  [vdso]                              [.] 0x00000000000006bb
   2,32%  [kernel]                            [k] __sched_text_start
   2,08%  libc-2.30.so                        [.] __sched_yield
   1,76%  openmsx-742-g64ef0812e              [.] openmsx::CPUCore<openmsx::R800TYPE>::executeInstructions
   1,67%  [kernel]                            [k] entry_SYSCALL_64_after_hwframe
   1,33%  openmsx-742-g64ef0812e              [.] openmsx::MSXMixer::generate
   1,25%  libc-2.30.so                        [.] __clock_gettime
   1,25%  openmsx-742-g64ef0812e              [.] openmsx::YM2413NukeYKT::YM2413::step18<false>
   1,22%  [vdso]                              [.] __vdso_clock_gettime
   1,15%  [kernel]                            [k] _nv030786rm
   1,13%  [kernel]                            [k] update_curr
   1,04%  [kernel]                            [k] _raw_spin_lock
   0,98%  perf_5.5                            [.] 0x0000000000273b33
   0,89%  [kernel]                            [k] native_sched_clock
   0,75%  [kernel]                            [k] cgroup_rstat_updated
   0,75%  libnvidia-glcore.so.440.82          [.] 0x0000000000ea4370
   0,73%  [kernel]                            [k] do_syscall_64
   0,72%  libc-2.30.so                        [.] __memcpy_ssse3
   0,70%  libnvidia-glcore.so.440.82          [.] 0x0000000000f9982a
   0,69%  libnvidia-glcore.so.440.82          [.] 0x0000000000f9a964
```
Looks like that commit has some nasty interaction with the nvidia driver?

**Comment by MBilderbeek** [comment]
OK, Wouter's theory was that the sync-to-vblank implementation is causing this (busy wait?) and indeed, when I set `__GL_SYNC_TO_VBLANK=0`, the CPU usage is back to normal.
So the real issue is that that commit somehow triggers sync to vblank... but why?
Any idea @D15C0DE ?

**Comment by mthuurne** [comment]
One thing that changed in that commit is that there is no longer a call to `SDL_CreateRenderer()` made in SDLGL-PP. Perhaps if we don't create an SDL renderer explicitly, SDL will create one implicitly with different flags? One of the renderer creation flags is `SDL_RENDERER_PRESENTVSYNC`, so if that is used for the default renderer, it might explain the difference.

In general I like vsynced page flipping, since it avoids tearing. However, if there are drivers that implement it by busy waiting then it's not really usable. Another issue is that the MSX frame rate might not match the host frame rate: often both are 60 Hz, but that isn't guaranteed to be the case, neither on the MSX nor on the host side. So it's probably safest to not sync to vblank, at least for now.

**Comment by D15C0DE** [comment]
It's quite plausible that that is the cause. I might have underestimated the possible side-effects of going with the defaults in my attempt to simplify the code as much as possible.

**Comment by MBilderbeek** [comment]
Should be fixed now with 9d85e52e731742c73fa3c5911a086ef985145e0e and now you can even change vsync yourself from within openMSX :)
